src/librustc_mir/interpret/eval_context.rs

```diff
@@ -1314,47 +1316,117 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
+            if in_range(value, size, scalar.valid_range.clone())? {
+                Ok(())
+            } else {
+                err!(ValidationFailure(format!(
+                    "encountered {:?} but expected something in the range {:?}",
+                    value,
+                    scalar.valid_range,
+                )))
+            }
+        }
+    }
+
     pub fn validate_ptr_target(
         &self,
         ptr: Pointer,
         ptr_align: Align,
-        ty: Ty<'tcx>
+        mut layout: TyLayout<'tcx>,
     ) -> EvalResult<'tcx> {
-        match ty.sty {
-            ty::TyBool => {
-                self.memory.read_scalar(ptr, ptr_align, Size::from_bytes(1))?.to_bool()?;
-            }
-            ty::TyChar => {
-                let c = self.memory.read_scalar(ptr, ptr_align, Size::from_bytes(4))?.to_bits(Size::from_bytes(4))? as u32;
-                match ::std::char::from_u32(c) {
-                    Some(..) => (),
-                    None => return err!(InvalidChar(c as u128)),
+        self.memory.dump_alloc(ptr.alloc_id);
+        trace!("validate_ptr_target: {:#?}", layout);
+        if let layout::Variants::Tagged { .. } = layout.variants {
+            let variant_index = self.read_discriminant_as_variant_index(
+                Place::from_ptr(ptr, ptr_align),
+                layout,
+            )?;
+            layout = layout.for_variant(self, variant_index);
+            trace!("variant layout: {:#?}", layout);
+        }
+        match layout.fields {
+            // primitives are unions with zero fields, because that's totally obvious and documented
+            layout::FieldPlacement::Union(0) => {
+                match layout.abi {
+                    // nothing to do, whatever the pointer points to, it is never going to be read
+                    layout::Abi::Uninhabited => Ok(()),
+                    // check that the scalar is a valid pointer or that its bit range matches the
+                    // expectation.
+                    layout::Abi::Scalar(ref scalar) => {
+                        let size = scalar.value.size(self);
+                        let value = self.memory.read_scalar(ptr, ptr_align, size)?;
+                        self.validate_scalar(value, size, scalar)
+                    },
+                    _ => bug!("bad abi for FieldPlacement::Union(0): {:#?}", layout.abi),
                 }
             }
-
-            ty::TyFnPtr(_) => {
-                self.memory.read_ptr_sized(ptr, ptr_align)?;
+            layout::FieldPlacement::Union(n) => {
+                // one of the fields needs to match, doesn't matter which, we leave that to
+                // the user. They are in unsafe territory anyway
+                for i in 0..n {
+                    let field_layout = layout.field(self, i)?;
+                    if self.validate_ptr_target(ptr, ptr_align, field_layout).is_ok() {
+                        return Ok(());
+                    }
+                }
```

Comment:
I don't think we make any validity assumptions for what is inside a union. But this is an open topic for discussion, and RFC-worthy. Saying "at least one field has to be valid" is probably conservative, but there should at least be a comment saying that no final decision has been made yet.

Reply:
I can turn this check into a lint that notes that we don't know if this is correct, pending an RFC

Reply:
I think having an error is okay, as it's forwarding-compatible to allow this later (right?). But either the error or the code should say that this is still subject of discussion.